blog/_posts/2025-04-04-Traffic-Routing-Demystified-Best-practices-for-AKS-Load-Balancers.md

```diff
@@ -0,0 +1,141 @@
+
+By combining these mechanisms, `externalTrafficPolicy=Local` provides a robust way to manage external traffic while maintaining source IP visibility and ensuring traffic is routed to healthy pods only.
+
+## Best Practices for Graceful Pod Shutdown
+
+Gracefully handling pod shutdowns is critical to maintaining service reliability and avoiding disruptions, especially in scenarios involving HTTP keep-alive connections or long-lived client sessions. Below are detailed best practices for how pods can handle Kubernetes initiated termination requests (eg: pod evictions or a scale down) to ensure a smooth shutdown process.
+
+### Gracefully Closing Existing Connections
+
+When a pod is shutting down (receiving the `TERM` signal), it is essential to ensure that existing client connections are closed properly to avoid abrupt disconnections or errors.
+
+- **For HTTP/1.1 Connections**:
+    The server should include a `Connection: close` header in its response for all active and new incoming requests. This informs clients not to reuse the connection and allows idle connections to be closed gracefully.
+    **Use Case**: Applications serving REST APIs or web traffic where clients rely on persistent connections for performance optimization.
+
+> **Note**: In HTTP/1.1, there is a potential race condition where the server might close an idle connection at the same time the client sends a new request. In such cases, the client must handle this scenario by retrying the request on a new connection.
+
+- **For HTTP/2 Connections**:
+    The server should send a `GOAWAY` frame to notify clients that the connection is being closed. This allows clients to gracefully terminate the connection and retry requests on a new connection if necessary.
+    **Use Case**: Applications using gRPC or HTTP/2 for high-performance communication between services or with external clients.
+
+### Preventing New Requests to an Unhealthy Pod
+To avoid routing new requests to a pod that is in the process of shutting down, it is important to manage its health status effectively. The below image shows the timeline for a pod receiving a `TERM` signal and gracefully shutting down without impact on external traffic:
+
+![Preventing New Requests to Unhealthy Pods](./AKS/blog/assets/images/preventingnewrequeststoanunhealthypod.png)
+
+- **Immediate Health Check Response**:
+    As soon as a pod is marked for deletion, its `healthCheckNodePort` should start returning a 500 error. This signals to external load balancers that the pod is no longer healthy and should not receive new traffic.
+
+- **Load Balancer Probe Delay**:
+    External load balancers may take a few seconds to detect the unhealthy status of a pod. During this time, the pod might still receive traffic.
```

Comment:
"may" - in fact it's "will" -- in the default AKS case it's a delay of up to around 10 seconds.

"receive traffic" - should be "receive new connections"